### PR TITLE
raft: raft_group0, register RPC verbs on all shards

### PR DESCRIPTION
**Pull request by gusev-p**
raft_group0 used to register RPC verbs only on shard 0. This worked on clusters with the same --smp setting on all nodes, since RPCs in this case are processed on the same shard as the calling code, and raft_group0 methods only run on shard 0.

A new test test_nodes_with_different_smp was added to identify the problem. Since --smp can
only be specified via the command line, a
corresponding parameter was added to the
ManagerClient.server_add method.
It allows to override the default parameters
set by the SCYLLA_CMDLINE_OPTIONS variable
by changing, adding or deleting individual
items.

Fixes: #12252

### CODE REVIEW

File: service/raft/raft_group0.cc
Comment (gleb-cloudius) [inline service/raft/raft_group0.cc:172]: There is `ser::group0_rpc_verb::unregister()` that unregister all verbs in the same idl. But this is pre-existing.

**Review by gleb-cloudius — APPROVED** [review]
Looks fine, but IMO better to separate test framework enhancement to a separate patch.

File: service/raft/raft_group0.cc
Comment (gusev-p) [reply service/raft/raft_group0.cc:172]: > but IMO better to separate test framework enhancement to a separate patch.

I think we've already discussed this, but I still don't feel comfortable with this approach of separating PRs into tiny patches.

Do you have specific, pragmatic reasons for doing it? With examples of situations where it pays off. Right now I see just the pain of separating these patches (fighting with git/github) and keeping them in sync, I don't see the point of it and it makes me frustrated.

> Patches should be self-contained and do one thing.

I see this in [Scylla guidelines](https://github.com/scylladb/scylladb/blob/master/docs/dev/review-checklist.md#patch-organization), but it doesn't specify what "self-contained" exactly means, nor does it give a rationale for why it matters. To me this patch is already self-contained, as it solves one specific problem and provides a test that demonstrates the problem. The code of the test and its associated helpers is just the code that needs to be written to solve this problem.

Another reason for not separating: having infra and examples of its usage in one place simplifies understanding for the reader, since he can see how parts of the infra are used and why they are important.

Comment (gleb-cloudius) [reply service/raft/raft_group0.cc:172]: > > but IMO better to separate test framework enhancement to a separate patch.
> 
> I think we've already discussed this, but I still don't feel comfortable with this approach of separating PRs into tiny patches.
> 
> Do you have specific, pragmatic reasons for doing it? With examples of situations where it pays off. Right now I see just the pain of separating these patches (fighting with git/github) and keeping them in sync, I don't see the point of it and it makes me frustrated.

Doing work is always a pin for the developer, but if it makes life easier for a reviewer it pays of. A patch that  does only one thing  is easier to review since a reviewer does not need to guess which change belongs to what fix. If a patch breaks something having  better separated changes makes bi-section easier. If it turned out that a fix needs to be backported cherry picking only minimal necessary change become possible.

If you want to look how development with "single patch per feature" work look at Cassandra where a single patch can rewrite the whole code base. Do not be like Cassandra.

Now I am not saying each patch should be no longer than X lines. If a single feature is large and isolated the patch implementing it can be large as well.

> 
> > Patches should be self-contained and do one thing.
> 
> I see this in [Scylla guidelines](https://github.com/scylladb/scylladb/blob/master/docs/dev/review-checklist.md#patch-organization), but it doesn't specify what "self-contained" exactly means, nor does it give a rationale for why it matters. To me this patch is already self-contained, as it solves one specific problem and provides a test that demonstrates the problem. The code of the test and its associated helpers is just the code that needs to be written to solve this problem.
>

It is hard to mathematically define self contained. But isn't it clear that changing a test framework has nothing to do with fixing the problem at hand? 
 
> Another reason for not separating: having infra and examples of its usage in one place simplifies understanding for the reader, since he can see how parts of the infra are used and why they are important.

Lest get this to the extreme: To linearise schema changes you need to have raft, so why not send the raft in the same single patch with the schema changes as well. According to you it *will* simplify the understanding, no?

Good explanation on infra change patch is what needed to simplify the understanding. Having an actual usage as a separate patch does not complicate anything. Tee code is there for everyone to look at.

Comment (gusev-p) [reply service/raft/raft_group0.cc:172]: OK, thanks. The strongest argument for me here is to make PR easier to review, I will stick with it while fighting with git.

Comment (gleb-cloudius) [reply service/raft/raft_group0.cc:172]: Do not underestimate backporting procedure needs. Suppose we have a another bug with different smp settings and the fix has a test case as well that uses the infra you added here. Unlike this one that other bug needs to be backported, but since it builds on this PR  a person who is doing the backport needs to split this PR manually. It is more work to him. Sometime a work he cannot do without deeper understanding of the problem.

Comment (gusev-p) [reply service/raft/raft_group0.cc:172]: That's understandable, but in the end, the tradeoff here is between the convenience of the PR's author and the convenience of everyone else. The increase in inconvenience for the author simply means he will have less motivation to refactor the code, make better infrastructure etc.

Comment (gleb-cloudius) [reply service/raft/raft_group0.cc:172]: There are always more developers than maintainers. Optimizations should be done to make life of maintainers easier. In the case of this PR it is trivial to split it using `git add -p`.

**Review by kostja — APPROVED** [review]

**Comment by gusev-p** [comment]
https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3600/

**Review by gleb-cloudius — APPROVED** [review]

File: service/raft/raft_group_registry.cc
Comment (gleb-cloudius) [inline service/raft/raft_group_registry.cc:198]: We are adding a continuation here even for the case where no copy is needed. I think we can do without this shared pointer. Something like that :
```
        return handle_raft_rpc(cinfo, gid, from, dst, [from, append_request = std::move(append_request), cpu = this_cpu()] (raft_rpc& rpc) mutable {
            if (this_cpu() != cpu) {
                raft::append_request r;
                r.current_term = append_request.current_term;
                r.prev...
                ....
                r.entries.reserve(r.entries.size());
                for (auto& e: r.entries) {
                    append_request.entries.push_back(make_lw_shared(*e));
                }
               rpc.append_entries(std::move(from), std::move(r));
            } else {
                rpc.append_entries(std::move(from), std::move(append_request));
            }
            return netw::messaging_service::no_wait();
    }
```

Comment (gleb-cloudius) [reply service/raft/raft_group_registry.cc:198]: Also for bisectability it is better to reorder last two patches.

Comment (gusev-p) [reply service/raft/raft_group_registry.cc:198]: > We are adding a continuation here even for the case where no copy is needed. I think we can do without this shared pointer. Something like that :
> 
> ```
>         return handle_raft_rpc(cinfo, gid, from, dst, [from, append_request = std::move(append_request), cpu = this_cpu()] (raft_rpc& rpc) mutable {
>             if (this_cpu() != cpu) {
>                 raft::append_request r;
>                 r.current_term = append_request.current_term;
>                 r.prev...
>                 ....
>                 r.entries.reserve(r.entries.size());
>                 for (auto& e: r.entries) {
>                     append_request.entries.push_back(make_lw_shared(*e));
>                 }
>                rpc.append_entries(std::move(from), std::move(r));
>             } else {
>                 rpc.append_entries(std::move(from), std::move(append_request));
>             }
>             return netw::messaging_service::no_wait();
>     }
> ```

this would imply decrementing the source lw_shared_ptr's counter on the target core in the destructor -> crash

Comment (gusev-p) [reply service/raft/raft_group_registry.cc:198]: > Also for bisectability it is better to reorder last two patches.

How specifically it is better?

Comment (gleb-cloudius) [reply service/raft/raft_group_registry.cc:198]: > > Also for bisectability it is better to reorder last two patches.
> 
> How specifically it is better?

It will be bisectable as opposite to not be bisectable in debug mode.

Comment (gleb-cloudius) [reply service/raft/raft_group_registry.cc:198]: > > We are adding a continuation here even for the case where no copy is needed. I think we can do without this shared pointer. Something like that :
> > ```
> >         return handle_raft_rpc(cinfo, gid, from, dst, [from, append_request = std::move(append_request), cpu = this_cpu()] (raft_rpc& rpc) mutable {
> >             if (this_cpu() != cpu) {
> >                 raft::append_request r;
> >                 r.current_term = append_request.current_term;
> >                 r.prev...
> >                 ....
> >                 r.entries.reserve(r.entries.size());
> >                 for (auto& e: r.entries) {
> >                     append_request.entries.push_back(make_lw_shared(*e));
> >                 }
> >                rpc.append_entries(std::move(from), std::move(r));
> >             } else {
> >                 rpc.append_entries(std::move(from), std::move(append_request));
> >             }
> >             return netw::messaging_service::no_wait();
> >     }
> > ```
> 
> this would imply decrementing the source lw_shared_ptr's counter on the target core in the destructor -> crash

Why? The lambda object, where original request is captured, is in the src shard, no?

Comment (gusev-p) [reply service/raft/raft_group_registry.cc]: You are right, I missed that, fixed.

Comment (gusev-p) [reply service/raft/raft_group_registry.cc]: > Also for bisectability it is better to reorder last two patches.

I don't understand, what bisectability means here.

Comment (gleb-cloudius) [reply service/raft/raft_group_registry.cc]: If each individual commit compiles and pass tests it is easier to do bisects if a bad commit does gets in (for instance the failure is rare). If the copy patch is after the registration patch there will be a test failure between them. If they are reordered there will not. Well, may be in this case this is not so important because we already have a failure, though different one...

Comment (gusev-p) [reply service/raft/raft_group_registry.cc]: OK

**Review by gleb-cloudius — APPROVED** [review]

File: test.py
Comment (kbr-scylla) [inline test.py:352]: You specified
```
        cmdline_from_test: List[str]
```
so it cannot be `None`, assuming we type-check the code.

I agree with the choice of `List[str]` instead of `Optional[List[str]]`, there's no point in passing `None`s when we can pass `[]`.

File: test/pylib/scylla_cluster.py
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:130]: Why is there a list of values for a key instead of a single value? What is the use case for this? In all examples you provided there at most one value for each key, so all lists you end up are of the form `[v]` where `v: str` or `[None]`. Wouldn't `Optional[str]` be simpler?

File: test/pylib/scylla_cluster.py
Comment (kbr-scylla) [inline test/pylib/scylla_cluster.py:118]: What's the point of `__missing__` though, if the same behavior can be obtained by simply not providing a value? (like the previous example)

File: raft/raft.hh
Comment (kbr-scylla) [inline raft/raft.hh:355]: I think this should be the copy constructor.

File: service/raft/raft_group0.cc
Comment (kbr-scylla) [inline service/raft/raft_group0.cc:147]: `() mutable`

File: service/raft/raft_group0.cc
Comment (kbr-scylla) [inline service/raft/raft_group0.cc:154]: `mutable`

File: test/topology/test_topology.py
Comment (kbr-scylla) [inline test/topology/test_topology.py:118]: That issue was already fixed and our build toolchain uses driver released with this fix.

File: test/topology/test_topology.py
Comment (kbr-scylla) [inline test/topology/test_topology.py:149]: How is this even possible, you're adding new nodes to a stopped cluster :O

File: test/topology/test_topology.py
Comment (kbr-scylla) [inline test/topology/test_topology.py:155]: Hm, but you're not restarting the servers that you stopped.
Whatever, the bigger problem is - why does this work at all, see my previous comment.

**Review by kbr-scylla — CHANGES_REQUESTED** [review]

File: raft/raft.hh
Comment (gleb-cloudius) [reply raft/raft.hh:355]: It cannot be the copy constructor since it makes a "deep" copy by copying the data the entries point too. Copy constructor would just copy shared pointers.

Comment (kbr-scylla) [reply raft/raft.hh:355]: The default copy constructor copies shared pointers, which is (arguably) incorrect, I'm suggesting to provide a custom copy constructor to make this deep copy.

Comment (kbr-scylla) [reply raft/raft.hh:355]: If we want to keep the default copy constructor for some reason, then:
- rename this function to `deep_copy()`,
- write a warning comment (here in the struct) that the copy constructor does a shallow copy and is unsafe e.g. for cross-shard situations.

Comment (gleb-cloudius) [reply raft/raft.hh:355]: > The default copy constructor copies shared pointers, which is (arguably) incorrect, I'm suggesting to provide a custom copy constructor to make this deep copy.

Why is it incorrect? It is correct for regular use. Cross shard copy is more of an exception.

Comment (kbr-scylla) [reply raft/raft.hh:355]: Deep copy is correct in all situations, shallow copy is a "dangerous" optimization that doesn't always work (cross-shard). I'd prefer we go with correctness by default, and optimizations should be explicitly requested when the caller guarantees they are safe, so I'd even go with having a deep copy in the copy constructor and a separate function (if needed) `shallow_copy()` for the optimized case. But meh, it's subjective.

Comment (gleb-cloudius) [reply raft/raft.hh:355]: When one uses shared pointer it explicitly says he wants shallow copy. If you do not want shallow copy just use the type directly. In C++ default copy constructor does deep copying. We use shared pointer because we want its behaviour in most cases. When we do not (as in this case) we add a function to do something different.

File: test/pylib/scylla_cluster.py
Comment (gusev-p) [reply test/pylib/scylla_cluster.py:145]: There is little value in playing a compiler role in the PR comments.

File: test/pylib/scylla_cluster.py
Comment (gusev-p) [reply test/pylib/scylla_cluster.py:130]: There is a specific example for this `# [--experimental-features, raft, --experimental-features, broadcast-tables]`.
In general, parameter can have many values.

File: test/pylib/scylla_cluster.py
Comment (gusev-p) [reply test/pylib/scylla_cluster.py:118]: The syntax is `--name name val`, we need explicit notion of parameter name and value to be able to merge two parameter sets. `__missing__` gives explicit sign that parameter doesn't have a value, otherwise the next word in the list would be treated as a value. The only case when we doesn't need this is at the end of the list, it's obvious that parameter doesn't have a value there.

File: raft/raft.hh
Comment (gusev-p) [reply raft/raft.hh:355]: I tried to entirely disable copy ctor/operator, but then aggregate initialisation required for RPC serialisation code stops working and I'm not particularly motivated about researching it. To me, all this matters is entirely subjective and shouldn't block the PR.

File: service/raft/raft_group0.cc
Comment (gusev-p) [reply service/raft/raft_group0.cc:147]: great point, thanks!

File: test/pylib/scylla_cluster.py
Comment (kbr-scylla) [reply test/pylib/scylla_cluster.py:145]: https://mypy-lang.org/

`pip install mypy`
then
`mypy <path-to-file>`

I recommend configuring your editor to use mypy in the background (I use vim + ALE and it works flawlessly)

Comment (kbr-scylla) [reply test/pylib/scylla_cluster.py:145]: I will eventually try to setup this in CI... Although @nyh might have a different opinion :P

Comment (kbr-scylla) [reply test/pylib/scylla_cluster.py:145]: But it should be possible to enforce `mypy` just in the places where we write type hints, it should ignore code without any type hints written. So it will still be optional, if someone doesn't want to use it won't be affected.

File: service/raft/raft_group0.cc
Comment (tgrabiec) [inline service/raft/raft_group0.cc:146]: Without RAII for the registration this is not exception safe. If registration fails during raft_group0::start(), already registered verbs will not be unregistered.

File: service/raft/raft_group0.cc
Comment (tgrabiec) [inline service/raft/raft_group0.cc:362]: Maybe uninit_rpc_verbs() should be static, since raft_group0 is not a sharded object. Calling an instance method from different shards on it is unsafe in general. Currently it only accesses sharded _ms, but making it static would make this safer.

File: test/pylib/scylla_cluster.py
Comment (kbr-scylla) [reply test/pylib/scylla_cluster.py:118]: > The syntax is `--name name val`

Not `--name val`? As in `--smp 2`.

> otherwise the next word in the list would be treated as a value

That depends on whether the next word starts with `-`.

Here: `[--key1, value, --key2]` you specify that `key2` has a missing value.
Here: `[--key1, --key2, value]` you specify that `key1` has a missing value. You know that the next word is not a value because it starts with `--`.

Perhaps it's just simpler to implement with `__missing__`?

File: service/raft/raft_group0.cc
Comment (gusev-p) [reply service/raft/raft_group0.cc]: made it static, along with `init_rpc_verbs`

File: test/pylib/manager_client.py
Comment (gusev-p) [reply test/pylib/manager_client.py:157]: fixed

File: test/topology/test_topology.py
Comment (kbr-scylla) [reply test/topology/test_topology.py:118]: > the driver persistently tries to connect to stopped node:

That's OK. It's not related to scylladb/python-driver#170 in any way.
Here you reconnect the driver basically to ignore the stopped nodes when creating a keyspace. scylladb/python-driver#170 was about driver not reconnecting after nodes were restarted (here you don't restart these stopped nodes).

> Anyway, maybe you know how to make it working in a more elegant way?

Just write an appropriate comment - explain why we need to reconnect the driver.

Anyway, this is part of a bigger story. I don't understand why the test works at all...

File: test/topology/test_topology.py
Comment (kbr-scylla) [reply test/topology/test_topology.py:149]: > The new leader should be elected.

You first have to add a node to the Raft configuration for it to be elected leader. You cannot add a node to Raft config without a quorum alive in the existing config. Here all existing nodes are dead. So you cannot even ask an existing node to start the reconfiguration procedure.

Even with Raft disabled, this should be impossible, a bootstrapping node needs to contact an existing node to bootstrap.

File: test/topology/test_topology.py
Comment (gusev-p) [reply test/topology/test_topology.py]: added a comment

We are starting a completely new, unrelated cluster, `_seeds` in `ScyllaCluster` ignores stopped nodes.

File: test/topology/test_topology.py
Comment (gusev-p) [reply test/topology/test_topology.py]: fixed the comment, the cluster is a fresh, new cluster unrelated to the previous one

File: test/topology/test_topology.py
Comment (gusev-p) [reply test/topology/test_topology.py:141]: This is a new, unrelated cluster, since `_seeds` in `ScyllaCluster` ignores stopped nodes

Comment (kbr-scylla) [reply test/topology/test_topology.py:141]: Ugh

File: test/topology/test_topology.py
Comment (kbr-scylla) [inline test/topology/test_topology.py:119]: So the test relies on a bug in the testing framework which allows starting two independent Scylla clusters under a single test framework cluster. (There should be 1-1 correspondence between framework's `ScyllaCluster` and an actual Scylla cluster). It basically works by accident.

Please modify the test so it doesn't depend on this bug.

Wouldn't it reproduce if:
- we didn't stop the existing servers,
- we added 2 new servers to the existing servers with --smp 3 and --smp 4; so in total we'd end up with 5 servers: 3 of them with --smp 2, one with --smp 3, and one with --smp 4. We have three different values of `smp` in this cluster, should be enough to reproduce the bug
- perform `add_tables` with this 5-node cluster. Maybe do it a few times to increase reproducibility.

Optionally we could decommission one of the original servers so that we have 4 servers in the cluster, forcing every quorum to have at least two different `smp`s. The decommission operation itself could reproduce the problem too.

Comment (kbr-scylla) [reply test/topology/test_topology.py:119]: Even adding the first server with a different `smp` may reproduce the problem, right?

Comment (gusev-p) [reply test/topology/test_topology.py:119]: I've spent some time and figured out the scenario where the problem reproduces in a most stable and natural way. The scenario - to deal with a fresh cluster with different `--smp` values. I do not want to adapt my testing needs to the nuances of the current testing infrastructure. This is a wrong path. It's a testing infra which should accommodate my needs. Otherwise, the tests will be dictated by the current infra and not the real world scenarios. 

The problem is that tests get the predefined cluster with existing nodes. Maybe it's what most tests want (I'm not sure), but it's not what is needed here. So I want to replace this cluster entirely. I can mark this action with some specific api in ScyllaCluster (`replace_cluster` or something) if that helps.

>Wouldn't it reproduce the targeted problem if:
> we didn't stop the existing servers,
> we added 2 new servers to the existing servers with --smp 3 and --smp 4; so in total we'd end up with 5 servers: 3 of them with --smp 2, one with --smp 3, and one with --smp 4. We have three different values of smp in this cluster, should be enough to reproduce the bug
> perform add_tables with this 5-node cluster. Maybe do it a few times to increase reproducibility.

I tried this scenario and various other, they were less stable to reproduce the problem.

Comment (kbr-scylla) [reply test/topology/test_topology.py:119]: > I do not want to adapt my testing needs to the nuances of the current testing infrastructure. This is a wrong path. It's a testing infra which should accommodate my needs. Otherwise, the tests will be dictated by the current infra and not the real world scenarios.

I agree that the framework is not flexible enough and needs to be extended. But it won't extend itself. Someone has to do it. Almost all tests in test_topology.py required extending the framework when they were written.

Right now, for a suite, there is a `Pool` of clusters used by all tests in the suite. The parameters of the cluster server are defined statically by the suite in a yaml file. This is quite limiting as you observed. For example, to test Raft upgrade scenarios I created a separate suite which starts clusters with Raft initially disabled.

It would be much better if each test case could specify the parameters of the cluster that it needs, and the pool would populate itself with clusters according to the needs of the collected test cases. More flexible, although it has downsides too, like less reusability of clusters between tests (although in topology tests currently we never reuse clusters because all become "dirty" after topology changes) and potentially much more complicated implementation of pool/test scheduler etc.

As a first step maybe we could allow tests to ask for empty clusters (using a special marker or fixture or fixture parameter?), so a test like this would have to start the cluster manually. Not sure how hard that would be to implement. May require a separate `Pool` of empty clusters or maybe we can mix the clusters in the existing `Pool` (probably bad idea, would lead to other tests obtaining these empty clusters as well).

> I tried this scenario and various other, they were less stable to reproduce the problem.

How frequently does the current approach reproduce the problem compared to e.g. what I proposed?

**Comment by kbr-scylla** [comment]
I'll merge this since it's an important bugfix, although I don't like the hacky way we test this with creating a separate cluster. The test is also unnecessarily long because it requires bootstrapping 6 nodes while it could work with 3 (the original 3 bootstrapped nodes are not used).

@gusev-p please create an issue for extending the test framework so we can bootstrap a single cluster with the desired parameters and fix the test.

**Comment by gusev-p** [comment]
> please create an issue for extending the test framework so we can bootstrap a single cluster with the desired parameters and fix the test.

https://github.com/scylladb/scylladb/issues/12442

File: service/raft/raft_group0.cc
Comment (tgrabiec) [reply service/raft/raft_group0.cc]: It may well be a preexisting problem, but we should follow up on this.

**Pull request closed without merging**
